content/blog/understanding-the-dom/index.md

```diff
@@ -0,0 +1,521 @@
+	- A node may have many children
+	- But may not have many parents
+- A non-root node may have many siblings as a result of the parent having many children
+
+![A chart showing the aforementioned rules of the node relationships](./dom_relationship_rules.svg)
+
+### How It's Used By The Browser {#how-the-browser-uses-the-dom}
+
+This tree tells the browser all of the information the browser needs to execute tasks in order to display and handle interaction with the user. For example, when the following CSS is applied to an `index.html` file:
+
+```css
+// index.css
+#b li {
+	background: red;
+}
+```
+
+```html
+<!-- index.html -->
+<main id="a">
+	<p id="e"></p>
+  <ul id="b">
+    <li id="c"></li>
+    <li id="d"></li>
+  </ul>
+</main>
+```
+
+The browser is able to keep in mind, while it's moving through the tree, that it needs to find an element with the `ID` of `b` and then mark it's `li` children with a red background. They're "children" because the DOM tree keeps that relationship info that's defined by the HTML.
+
+![A chart showing the 'ul' tag highlighted in green with the children 'li' tags marked in red](./dom_tree_with_css.svg "Diagram showing the above code as a graph")
+
+> The `ul` element is marked as green just to showcase that it is the element being marked by the first part of the selector
+
+Typically, the browser will "visit" it's nodes in a specific order. For example, in the above chart, the browser might start at the `main` tag, then go to the `p` tag, then visit the `ul` tag, and finally the two children in order from left-to-right (`li id="c"` , `li id="d"`).
+
+The browser, knowing what CSS to look for, is able to see the `ul` with the correct ID and know to mark it's children with the correct metadata that matches the selector with the relevant CSS.
+
+This tree relationship also enables CSS selectors such as [general sibling selector (`~` )](https://developer.mozilla.org/en-US/docs/Web/CSS/General_sibling_combinator) or the [adjacent sibling selector (`+`)](https://developer.mozilla.org/en-US/docs/Web/CSS/Adjacent_sibling_combinator) to find siblings to a given selector
+
+![A showcase of the above selectors and how they always look forward, never behind](css_selectors_demo.svg)
+
+> Interestingly, one of the questions that I've often heard asked is a "parent selector". The idea behind this is that the [direct child selector (`>`)](https://developer.mozilla.org/en-US/docs/Web/CSS/Child_combinator)  exists, why not have the ability to mark any parent of `.classname` selectors. 
+>
+> The answer behind that is: Performance. The [W3 organization](https://www.w3.org/Style/CSS/#specs) (the organization who maintains the HTML and CSS standard specification) points to the tree structure of the DOM and the algorithm behind how browsers traverse the DOM (or, "visit" the nodes in order to figure out what CSS to apply)  as not being performant when allowing parent selectors.
```

Comment:
```suggestion
> The answer behind that is performance. The [W3 organization](https://www.w3.org/Style/CSS/#specs) (the organization who maintains the HTML and CSS standard specification) points to the tree structure of the DOM and the algorithm behind how browsers traverse the DOM (or, "visit" the nodes in order to figure out what CSS to apply)  as not being performant when allowing parent selectors.
```
- `is` is functionally equivalent to the colon here
  - You can also replace `is` with the colon (and lower casing `Performance`) or use an ellipsis if the pause is important

Reply:
```suggestion
> The answer behind that: performance. The [W3 organization](https://www.w3.org/Style/CSS/#specs) (the organization who maintains the HTML and CSS standard specification) points to the tree structure of the DOM and the algorithm behind how browsers traverse the DOM (or, "visit" the nodes in order to figure out what CSS to apply)  as not being performant when allowing parent selectors.
```